src/common/io.py

```diff
@@ -1373,10 +1373,12 @@ def _get_diagnostics_trajectory(self, data_index, start_index = 0, stop_index =
         """ Returns trajectory for the diagnostics variable that corresponds to index 'data_index'. """
         self._verify_file_data()
 
-        if data_index in self._data_3:
+        index_in_cache = data_index in self._data_3
+        partial_cache_ok = self._can_use_partial_cache(start_index, stop_index)
+        if (index_in_cache and not self._allow_file_updates) or (index_in_cache and partial_cache_ok):
```

Comment:
Now caching also works as expected with start/stop, including the diagnostics variables.